File: pacotes/protocolo/src/codificador.ts
import { calcularCRC16 } from './crc16.js';
import {
  MAGIC, VERSAO_PROTO,
  TIPO_DADOS, TIPO_CONFIGURACAO, TIPO_STATUS,
  CMD_TARAR, CMD_CALIBRAR, CMD_OBTER_CONFIG, CMD_DEFINIR_PARAM,
  type PacoteESP32, type ComandoHost,
  type PacoteDados, type PacoteConfiguracao, type PacoteStatus,
  type ComandoTarar, type ComandoCalibar, type ComandoObterConfig, type ComandoDefinirParam,
} from './tipos.js';

// Tamanhos fixos dos pacotes (bytes)
const TAM_DADOS    = 20;
const TAM_CONFIG   = 64;
const TAM_STATUS   = 14;
const TAM_CMD_TARA  = 8;
const TAM_CMD_CALIB = 10;
const TAM_CMD_GET   = 8;
const TAM_CMD_SET   = 18;

function view(size: number): [DataView, Uint8Array] {
  const buf = new ArrayBuffer(size);
  return [new DataView(buf), new Uint8Array(buf)];
}

function assinarCRC(bytes: Uint8Array): void {
  const crc = calcularCRC16(bytes.subarray(0, bytes.length - 2));
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  view.setUint16(bytes.length - 2, crc, true); // little-endian
}

function verificarCRC(bytes: Uint8Array): void {
  const crcRecebido = new DataView(bytes.buffer, bytes.byteOffset).getUint16(bytes.length - 2, true);
  const crcCalculado = calcularCRC16(bytes.subarray(0, bytes.length - 2));
  if (crcCalculado !== crcRecebido) {
    throw new Error(`CRC inválido: recebido=0x${crcRecebido.toString(16)} calculado=0x${crcCalculado.toString(16)}`);
  }
}

// ─── Codificadores (objeto → bytes) ─────────────────────────────────────────

function codificarComandoTarar(_cmd: ComandoTarar): Uint8Array {
  const [dv, bytes] = view(TAM_CMD_TARA);
  dv.setUint16(0, MAGIC, true);
  dv.setUint8(2, VERSAO_PROTO);
  dv.setUint8(3, CMD_TARAR);
  dv.setUint16(4, 0, true);
  assinarCRC(bytes);
  return bytes;
}

function codificarComandoCalibar(cmd: ComandoCalibar): Uint8Array {
  const [dv, bytes] = view(TAM_CMD_CALIB);
  dv.setUint16(0, MAGIC, true);
  dv.setUint8(2, VERSAO_PROTO);
  dv.setUint8(3, CMD_CALIBRAR);
  dv.setFloat32(4, cmd.massaG, true);
  assinarCRC(bytes);
  return bytes;
}

function codificarComandoObterConfig(_cmd: ComandoObterConfig): Uint8Array {
  const [dv, bytes] = view(TAM_CMD_GET);
  dv.setUint16(0, MAGIC, true);
  dv.setUint8(2, VERSAO_PROTO);
  dv.setUint8(3, CMD_OBTER_CONFIG);
  dv.setUint16(4, 0, true);
  assinarCRC(bytes);
  return bytes;
}

function codificarComandoDefinirParam(cmd: ComandoDefinirParam): Uint8Array {
  const [dv, bytes] = view(TAM_CMD_SET);
  dv.setUint16(0, MAGIC, true);
  dv.setUint8(2, VERSAO_PROTO);
  dv.setUint8(3, CMD_DEFINIR_PARAM);
  dv.setUint8(4, cmd.paramId);
  // bytes 5,6,7 = reserved (zero)
  dv.setFloat32(8, cmd.valorF, true);
  dv.setUint32(12, cmd.valorI >>> 0, true);
  assinarCRC(bytes);
  return bytes;
}

/**
 * Codifica um ComandoHost em bytes prontos para enviar ao ESP32.
 */
export function codificarComando(cmd: ComandoHost): Uint8Array {
  switch (cmd.tipo) {
    case 'CMD_TARAR':         return codificarComandoTarar(cmd);
    case 'CMD_CALIBRAR':      return codificarComandoCalibar(cmd);
    case 'CMD_OBTER_CONFIG':  return codificarComandoObterConfig(cmd);
    case 'CMD_DEFINIR_PARAM': return codificarComandoDefinirParam(cmd);
  }
}

// ─── Decodificadores (bytes → objeto) ───────────────────────────────────────

function decodificarDados(bytes: Uint8Array): PacoteDados {
  verificarCRC(bytes);
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  return {
    tipo:           'DADOS',
    marcaTemporal:  dv.getUint32(4, true),
    forcaNewtons:   dv.getFloat32(8, true),
    forcaBruta:     dv.getInt32(12, true),
    statusFirmware: dv.getUint8(16),
  };
}

function decodificarConfiguracao(bytes: Uint8Array): PacoteConfiguracao {
  verificarCRC(bytes);
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  let off = 4;
  const fatorConversao      = dv.getFloat32(off, true); off += 4;
  const gravidade           = dv.getFloat32(off, true); off += 4;
  const leiturasEstaveis    = dv.getUint16(off, true);  off += 2;
  const toleranciaEst       = dv.getFloat32(off, true); off += 4;
  const numAmostrasMedia    = dv.getUint16(off, true);  off += 2;
  const numAmostrasCal      = dv.getUint16(off, true);  off += 2;
  const usarMediaMovel      = dv.getUint8(off) !== 0;   off += 1;
  const usarEMA             = dv.getUint8(off) !== 0;   off += 1;
  const timeoutCal          = dv.getUint16(off, true);  off += 2;
  const offsetTara          = dv.getInt32(off, true);   off += 4;
  const capacidadeMaxGramas = dv.getFloat32(off, true); off += 4;
  const acuracia            = dv.getFloat32(off, true); off += 4;
  const modo                = dv.getUint8(off);

  return {
    tipo: 'CONFIGURACAO',
    fatorConversao, gravidade, leiturasEstaveis, toleranciaEst,
    numAmostrasMedia, numAmostrasCal, usarMediaMovel, usarEMA,
    timeoutCal, offsetTara, capacidadeMaxGramas, acuracia, modo,
  };
}

function decodificarStatus(bytes: Uint8Array): PacoteStatus {
  verificarCRC(bytes);
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  return {
    tipo:          'STATUS',
    tipoStatus:    dv.getUint8(4),
    codigo:        dv.getUint8(5),
    valor:         dv.getUint16(6, true),
    marcaTemporal: dv.getUint32(8, true),
  };
}

/**
 * Decodifica um buffer bruto (do ESP32) em um PacoteESP32 tipado.
 * Lança Error se o buffer for inválido, truncado ou com CRC errado.
 */
export function decodificar(bytes: Uint8Array): PacoteESP32 {
  if (bytes.length < 4) {
    throw new Error('Buffer muito curto');
  }

  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const magic = dv.getUint16(0, true);
  if (magic !== MAGIC) {
    throw new Error(`Magic inválido: 0x${magic.toString(16)}`);
  }

  const tipo = dv.getUint8(3);

  const tamEsperado: Record<number, number> = {
    [TIPO_DADOS]:        TAM_DADOS,
    [TIPO_CONFIGURACAO]: TAM_CONFIG,
    [TIPO_STATUS]:       TAM_STATUS,
  };

  const tam = tamEsperado[tipo];
  if (tam === undefined) {
    throw new Error(`Tipo desconhecido: 0x${tipo.toString(16)}`);
  }
  if (bytes.length < tam) {
    throw new Error(`Buffer muito curto para tipo 0x${tipo.toString(16)}: esperado ${tam}, recebido ${bytes.length}`);
  }

  const pacoteBytes = bytes.subarray(0, tam);

  switch (tipo) {
    case TIPO_DADOS:        return decodificarDados(pacoteBytes);
    case TIPO_CONFIGURACAO: return decodificarConfiguracao(pacoteBytes);
    case TIPO_STATUS:       return decodificarStatus(pacoteBytes);
    default:
      throw new Error(`Tipo desconhecido: 0x${tipo.toString(16)}`);
  }
}

export { calcularCRC16 } from './crc16.js';
export * from './tipos.js';

File: android/app/src/main/java/br/edu/ifsc/balancagfig/protocolo/Codificador.kt
package br.edu.ifsc.balancagfig.protocolo

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Codec binário ESP ↔ Host. Espelho de pacotes/protocolo/src/codificador.ts:
 * todos os campos em little-endian, CRC16 nos dois últimos bytes.
 */
object Codificador {

    // ─── Codificadores (comando → bytes) ────────────────────────────────────

    fun codificar(cmd: ComandoHost): ByteArray = when (cmd) {
        is ComandoTarar -> cabecalho(Protocolo.TAM_CMD_TARA, Protocolo.CMD_TARAR)
            .putShort(0)
            .assinar()
        is ComandoCalibrar -> cabecalho(Protocolo.TAM_CMD_CALIB, Protocolo.CMD_CALIBRAR)
            .putFloat(cmd.massaG)
            .assinar()
        is ComandoObterConfig -> cabecalho(Protocolo.TAM_CMD_GET, Protocolo.CMD_OBTER_CONFIG)
            .putShort(0)
            .assinar()
        is ComandoDefinirParam -> cabecalho(Protocolo.TAM_CMD_SET, Protocolo.CMD_DEFINIR_PARAM)
            .put(cmd.paramId.toByte())
            .put(0).put(0).put(0)                    // bytes 5,6,7 reservados
            .putFloat(cmd.valorF)
            .putInt(cmd.valorI.toInt())              // uint32
            .assinar()
    }

    private fun cabecalho(tamanho: Int, tipo: Int): ByteBuffer =
        ByteBuffer.allocate(tamanho).order(ByteOrder.LITTLE_ENDIAN)
            .putShort(Protocolo.MAGIC.toShort())
            .put(Protocolo.VERSAO.toByte())
            .put(tipo.toByte())

    /** Grava o CRC dos bytes anteriores nos 2 últimos bytes e devolve o array. */
    private fun ByteBuffer.assinar(): ByteArray {
        val bytes = array()
        val crc = Crc16.calcular(bytes, 0, bytes.size - 2)
        putShort(bytes.size - 2, crc.toShort())
        return bytes
    }

    // ─── Decodificadores (bytes → pacote) ───────────────────────────────────

    /**
     * Decodifica um buffer bruto do ESP. Lança [ErroProtocolo] se o buffer for
     * inválido, truncado ou com CRC errado. Bytes excedentes são ignorados.
     */
    fun decodificar(bytes: ByteArray): PacoteESP {
        if (bytes.size < 4) throw ErroProtocolo("Buffer muito curto")

        val bb = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
        val magic = bb.getShort(0).toInt() and 0xffff
        if (magic != Protocolo.MAGIC) {
            throw ErroProtocolo("Magic inválido: 0x${magic.toString(16)}")
        }

        val tipo = bb.get(3).toInt() and 0xff
        val tam = Protocolo.tamanhoPorTipo(tipo)
            ?: throw ErroProtocolo("Tipo desconhecido: 0x${tipo.toString(16)}")
        if (bytes.size < tam) {
            throw ErroProtocolo(
                "Buffer muito curto para tipo 0x${tipo.toString(16)}: esperado $tam, recebido ${bytes.size}"
            )
        }

        verificarCrc(bytes, tam)

        return when (tipo) {
            Protocolo.TIPO_DADOS -> PacoteDados(
                marcaTemporal = bb.getInt(4).toLong() and 0xffffffffL,
                forcaNewtons = bb.getFloat(8),
                forcaBruta = bb.getInt(12),
                statusFirmware = bb.get(16).toInt() and 0xff,
            )
            Protocolo.TIPO_CONFIGURACAO -> PacoteConfiguracao(
                fatorConversao = bb.getFloat(4),
                gravidade = bb.getFloat(8),
                leiturasEstaveis = bb.getShort(12).toInt() and 0xffff,
                toleranciaEst = bb.getFloat(14),
                numAmostrasMedia = bb.getShort(18).toInt() and 0xffff,
                numAmostrasCal = bb.getShort(20).toInt() and 0xffff,
                usarMediaMovel = bb.get(22).toInt() != 0,
                usarEMA = bb.get(23).toInt() != 0,
                timeoutCal = bb.getShort(24).toInt() and 0xffff,
                offsetTara = bb.getInt(26),
                capacidadeMaxGramas = bb.getFloat(30),
                acuracia = bb.getFloat(34),
                modo = bb.get(38).toInt() and 0xff,
            )
            else -> PacoteStatus(
                tipoStatus = bb.get(4).toInt() and 0xff,
                codigo = bb.get(5).toInt() and 0xff,
                valor = bb.getShort(6).toInt() and 0xffff,
                marcaTemporal = bb.getInt(8).toLong() and 0xffffffffL,
            )
        }
    }

    private fun verificarCrc(bytes: ByteArray, tam: Int) {
        val bb = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
        val recebido = bb.getShort(tam - 2).toInt() and 0xffff
        val calculado = Crc16.calcular(bytes, 0, tam - 2)
        if (recebido != calculado) {
            throw ErroProtocolo(
                "CRC inválido: recebido=0x${recebido.toString(16)} calculado=0x${calculado.toString(16)}"
            )
        }
    }
}
